### DISCUSSION TITLE
Avalonia port

### DISCUSSION THREAD

**Discussion by Mrxx99** [Tyrrrz/YoutubeDownloader#295] [General]
@Tyrrrz  I started porting but have some questions.
Would you want to have a separate project (YoutubeDownloader.Avalonia) or directly replacement?
If separate  project it would have much duplicated code (ViewModels, Services)

For the libraries that are used:
- MaterialDesign*: there is a port available (https://github.com/AvaloniaCommunity/Material.Avalonia)
- Microsoft.Xaml.Behaviors.Wpf: there is a port available (https://github.com/wieslawsoltes/AvaloniaBehaviors)
- Ookii.Dialogs.Wpf: that is not needed anymore, Avalonia has access to native folder/file dialogs
- PropertyChanged.Fody: while Fody is not WPF specific it has issues with Avalonia, it should be removed
- Stylet: this is WPF only, it needs an replacement
The other libraries should just work

For MVVM I would suggest the CommunityToolkit.MVVM with Microsoft.Extensions.DependencyInjection would that be ok?

The ffmpeg.exe is only for Windows, do we need a replacement for Linux and Mac?
(I can only test Windows and Linux, I don't have a Mac)

**Comment by Tyrrrz** [comment]
Hi @Mrxx99 

> Would you want to have a separate project (YoutubeDownloader.Avalonia) or directly replacement?

A direct replacement would be better I think.

> PropertyChanged.Fody: while Fody is not WPF specific it has issues with Avalonia, it should be removed

Is there anything like `PropertyChanged.Fody` for Avalonia? I would hate to go back to writing manual properties with INPC triggers :/

> For MVVM I would suggest the CommunityToolkit.MVVM with Microsoft.Extensions.DependencyInjection would that be ok?

Yeah, I'm fine with whatever.

> The ffmpeg.exe is only for Windows, do we need a replacement for Linux and Mac?

Yeah there's a build for each system. I think the easiest approach is to have YoutubeDownloader download the FFmpeg binary itself on start (if it's missing).

https://ffmpeg.org/download.html

**Reply by Mrxx99** [reply]
> Is there anything like PropertyChanged.Fody for Avalonia? I would hate to go back to writing manual properties with INPC triggers :/

There is [ReactiveUI.Fody](https://www.reactiveui.net/docs/handbook/view-models/boilerplate-code), where you only need an attribute on each property. But sadly it only works with base classes implementing ReactiveObject (a ViewModel base), so it won't work for SettingsService, but for everything else.

**Reply by Tyrrrz** [reply]
That's quite sad. The biggest boon of `PropertyChanged.Fody` is how it also wires change handlers for evaluated properties, it would be really sad to lose it.

Is there a link that details what issues it has with Avalonia? Maybe we can fix/work around them?

**Reply by Mrxx99** [reply]
It still works with delegating to properties like in the SettingsViewModel, only for the SettingsService it would not work (because it uses SettingsManager as base class).

The issue is here: https://github.com/Fody/PropertyChanged/issues/335
It was closed with the note it won't be supported for Avalonia OOTB, but I did not yet find a solution.
It seems Avalonia would have to make changes to it's INotifyPropertyChanged implementation of the controls base classes (AvaloniaObject) for it to work.

**Reply by Mrxx99** [reply]
ok, one workaround is the [DoNotNotify] attribute on every view, I somehow overlooked that.

**Reply by Tyrrrz** [reply]
Or we can have a custom base class for views maybe, that has a `RaisePropertyChanged` with supported signature? Or just put the `[DoNotNotify]` on the base class.

**Reply by Mrxx99** [reply]
Yes, that would be the easiest, so we would only need the attribute at two places, the UserControlBase class and the Window.

Also for Styles there is this discussion on Avalonia support: https://github.com/canton7/Stylet/discussions/342
And a started proof of concept (without a nuget package): https://github.com/alexhelms/StyletAvalonia
There is also a chinese port in preview with a nuget package (requires also preview of Avalonia): https://github.com/rsdte/Stylet.Avalonia

**Reply by Tyrrrz** [reply]
I'm okay with dropping Stylet as it doesn't do anything special. We can replace with any framework you prefer. Ideally something more or less conventional in terms of MVVM, but I'm okay with magic if it saves times and has good documentation. For example, not having to create commands and have bindings directly to methods is nice, but I vaguely remember Avalonia supporting this OOTB?

**Reply by Mrxx99** [reply]
Yes, Avalonia supports binding to Methods OOTB

**Reply by Tyrrrz** [reply]
Maybe we won't need an MVVM framework at all with Avalonia? I.e. just a dependency container and that's it. I assume base INPC implementation is also built-in.

**Comment by Mrxx99** [comment]
@Tyrrrz Small update: I am still working on it. Most already works, still working on some details.
I am currently using Avalonia 11-preview4 the full release of v11 should come sometime this year (seems like Q1), I would plan to finish it  after that release.
As Onova does only support Windows, how would you like to handle other platforms? The easiest would be to have no auto-updater (at least for start) for other platforms than Windows. Or should I look into other solutions like [NetSparke](https://github.com/NetSparkleUpdater/NetSparkle)?
The DotnetRuntimeBootstrapper also only works for Windows, so I guess for other platforms it should be self-contained instead?
I had to change the startup check from a ModuleInitializer to checking during startup, that shows a blocking dialog above the main window (could maybe also changed to overlay popup like the other dialogs) , because the Avalonia.Messagebox does need the Avalonia app to be initialized first.

**Reply by Tyrrrz** [reply]
Hi @Mrxx99! That's cool, no worries.

> As Onova does only support Windows, how would you like to handle other platforms? The easiest would be to have no auto-updater (at least for start) for other platforms than Windows. Or should I look into other solutions like [NetSparke](https://github.com/NetSparkleUpdater/NetSparkle)?

To keep it simple, let's only use Onova for Windows and not provide any auto-update solution for other platforms yet.

> The DotnetRuntimeBootstrapper also only works for Windows, so I guess for other platforms it should be self-contained instead?

Let's keep the framework-dependent approach for other platforms as well, to make the initial move to Avalonia simpler. We can change it later in the future.

> I had to change the startup check from a ModuleInitializer to checking during startup, that shows a blocking dialog above the main window (could maybe also changed to overlay popup like the other dialogs) , because the Avalonia.Messagebox does need the Avalonia app to be initialized first.

Okay, you can remove that part from your PR, I was going to rework it into a separate cross-platform NuGet later anyway.
